nats.go

```diff
@@ -0,0 +1,245 @@
+package gorillaz
+
+import (
+	"context"
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/opentracing/opentracing-go"
+	"github.com/skysoft-atm/gorillaz/stream"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+	"time"
+)
+
+// PullNatsStream returns the next subject and event for the given stream and consumer
+// If the event is processed successfully, it must be acknowledge to get a new message. If the message is not acknowledge, then PullNatsStream will return the same event multiple times
+// If no message is available after the ctx timeout, then an error nats.ErrTimeout is returned with an empty subject and an event nil
+func (g *Gaz) PullNatsStream(ctx context.Context, stream string, consumer string) (subject string, event *stream.Event, err error) {
+	subj := "$JS.API.CONSUMER.MSG.NEXT." + stream + "." + consumer
+	msg, err := g.NatsConn.RequestWithContext(ctx, subj, nil)
+	if err != nil {
+		return "", nil, err
+	}
+	e := msgToEvent(msg)
+	e.AckFunc = func() error {
+		return msg.Respond(nil)
+	}
+	return msg.Subject, &e, nil
+}
+
+// MsgHandler handles received events from Nats
+// If NatsConsumerOpts.AutoAck is set, if MsgHandler returns no error, the message will be acknowledged. If an error is returned, the event won't be acknowledged.
+type MsgHandler func(subject string, event stream.Event) (reply *stream.Event, err error)
+
+type NatsConsumerOpts struct {
+	AutoAck bool
+	Queue   string
+}
+
+type NatsConsumerOpt func(n *NatsConsumerOpts)
+
+// WithAutoAck automatic acknowledge message received if MsgHandle returns no error
+func WithAutoAck() NatsConsumerOpt {
+	return func(o *NatsConsumerOpts) {
+		o.AutoAck = true
+	}
+}
+
+// WithQueue configures Nats Queue consumer
+func WithQueue(queue string) NatsConsumerOpt {
+	return func(o *NatsConsumerOpts) {
+		o.Queue = queue
+	}
+}
+
+// SubscribeNatsSubject subscribes to a Nats stream, and forward received messages to handler
+// An error is returned if the subscription fails, but not when the connection with Nats is interrupted
+func (g *Gaz) SubscribeNatsSubject(subject string, handler MsgHandler, opts ...NatsConsumerOpt) (*NatsSubscription, error) {
+	c := &NatsConsumerOpts{
+		AutoAck: false,
+	}
+	for _, o := range opts {
+		o(c)
+	}
+	if g.NatsConn == nil {
+		return nil, fmt.Errorf("gorillaz nats connection is nil, cannot consume stream")
+	}
+
+	do := func(m *nats.Msg) {
+		e := msgToEvent(m)
+
+		if !c.AutoAck && m.Reply != "" {
+			e.AckFunc = func() error {
+				return m.Respond(nil)
+			}
+		}
+
+		reply, err := handler(m.Subject, e)
+		if err == nil {
```

Comment:
the error case is not handled

Reply:
it is actually, in case of error, if there is a response, it's sent as a response